contracts/Market.sol

```diff
@@ -1610,6 +1369,89 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts, suppliers);
   }
 
+  /**
+   * @notice Allocates removals from a specific supplier to be fulfilled.
+   * @dev This function is responsible for validating and allocating the supply from a specific supplier.
+   * @param certificateAmount The total amount of NRTs for the certificate.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction.
+   * @param vintages A set of valid vintages from which to allocate removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateRemovalsFromSupplierSpecificVintages(
+    uint256 certificateAmount,
+    address supplier,
+    uint256[] memory vintages
+  )
+    internal
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    (
+      countOfRemovalsAllocated,
+      ids,
+      amounts
+    ) = _allocateSupplySingleSupplierSpecificVintages({
+      amount: certificateAmount,
+      supplier: supplier,
+      vintages: vintages
+    });
+    suppliers = new address[](countOfRemovalsAllocated).fill({value: supplier});
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Allocates removals to fulfill an order from the specified vintages.
+   * @dev This function is responsible for validating and allocating the supply to fulfill an order.
+   * @param purchaser The address of the purchaser.
+   * @param certificateAmount The total amount for the certificate.
+   * @param vintages A set of valid vintages from which to allocate removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateRemovalsSpecificVintages(
+    address purchaser,
+    uint256 certificateAmount,
+    uint256[] memory vintages
+  )
+    internal
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 availableSupply = _removal.getMarketBalance();
+    _validateSupply({
```

Comment:
I added tests for the cases both in which there is no such supplier in the market, and where the specified vintages don't exist.  They both revert with `InsufficientSupply`.  The reason I don't think we should add validation before the actual fulfillment loop is that we will repeat work.  In both of these cases, we learn quite early in the fulfillment process whether the supplier has removals and/or whether there are removals available in the requested vintages.  For example, if there is no supplier or none of the requested vintages, we basically have some loops that never run (because they're iterating over 0 things) and then we conclude insufficient supply after doing just as much work as we would have to do to make this determination ahead of the fulfillment algorithm. A check ahead of time would simply duplicate the work of looking into those data structures immediately before doing it anyway.